drivers/source/UnbufferedSerial.cpp

```diff
@@ -0,0 +1,105 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/UnbufferedSerial.h"
+
+#if DEVICE_SERIAL
+
+#include "platform/mbed_critical.h"
+
+namespace mbed {
+
+UnbufferedSerial::UnbufferedSerial(
+    PinName tx,
+    PinName rx,
+    int baud
+) : SerialBase(tx, rx, baud)
+{
+    // No lock needed in the constructor
+}
+
+ssize_t UnbufferedSerial::write(const void *buffer, size_t size)
+{
+    const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    bool lock_api = !core_util_in_critical_section();
+
+    if (lock_api) {
+        lock();
+    }
+
+    for (size_t i = 0; i < size; i++) {
+        _base_putc(buf[i]);
+    }
+
+    if (lock_api) {
+        unlock();
+    }
+
+    return size;
+}
+
+ssize_t UnbufferedSerial::read(void *buffer, size_t size)
+{
+    unsigned char *buf = static_cast<unsigned char *>(buffer);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    lock();
+
+    buf[0] = _base_getc();
+
+    unlock();
+
+    return 1;
+}
+
+short UnbufferedSerial::poll(short events)
+{
+    short revents = 0;
+    if ((events & POLLIN) && SerialBase::readable()) {
+        revents |= POLLIN;
+    }
+    if ((events & POLLOUT) && SerialBase::writeable()) {
+        revents |= POLLOUT;
+    }
+    return revents;
+}
+
+/** Acquire exclusive access to this serial port
+ */
+void UnbufferedSerial::lock()
+{
+    // No lock used - external synchronization required
```

Comment:
Why do we want to override Serial::lock?

Reply:
Overriding it means it is no longer dependent of `SerialBase`'s implementation. It is guaranteed to always do nothing regardless on what happens to `SerialBase` implementation in the future. The same is done in `RawSerial` and `UARTSerial`.

Reply:
Actually, this is a bit weird.

I thought the override was there because `SerialBase` didn't actually have an implementation of the methods.

It doesn't really matter much, but would save a teeny amount of ROM if you were mixing `BufferedSerial` and `UnbufferedSerial` to have them both use `SerialBase`'s common "do nothing", rather than each have their own "do nothing".

I'd be in favour of removing the definitions here, and doing the same when you make `BufferedSerial`. We're totally dependent on `SerialBase`'s implementation for everything else, so not sure why we shouldn't depend on this.

Reply:
[This](https://github.com/ARMmbed/mbed-os/pull/11961#discussion_r353014835) is the only explanation I could come up with to explain it as it is also done for those two other classes. If we want to rely on `SerialBase` then I will remove it.

Reply:
I don't think there was a real reason for it (perhaps there was a different implementation at some time and it was then removed?). So happy with this code removed in the latest version.